src/datasets/arrow_dataset.py

```diff
@@ -420,6 +420,26 @@ def to_tf_dataset(
             batch_size=batch_size if drop_remainder else None,
         )
 
+        shape_verification_signature, _ = dataset._get_output_signature(
```

Comment:
why do you need to call it a second time ? can't this logic be inside _get_output_signature ?

Reply:
That would make sense, actually! I'll move it.

Reply:
@lhoestq I cleaned things up a lot based on your feedback - `_get_output_signature` is only called once, and it now immediately samples 200 batches of size 2 to infer the shape, but then overwrites the batch size element of the inferred shape with the actual batch size.

Reply:
Cool! :)

I also think 10 batches is good by default, going to 200 batches can take too much time for some datasets IMO

Reply:
I actually specifically had problems with incorrect inferences when using 10! I think it's preferable for `to_tf_dataset()` to be a little slow sometimes (it's only called once at dataset creation time) than to infer wrong shapes and create tricky bugs for users.

If you want, though, I can make `num_test_batches` an argument to `to_tf_dataset`?

Reply:
Gentle ping @lhoestq!

Reply:
Pinging @albertvillanova instead because Quentin is on holiday and I didn't realize >.>;

Reply:
> I actually specifically had problems with incorrect inferences when using 10!

Can you explain what problems ?

Reply:
In some cases, sampling 10 batches from the dataset makes it look like the dataset has a constant shape, but actually it doesn't. This is particularly common when datasets have been truncated. For example, if the average length in a dataset before truncation is >> 512, but we truncate at 512, then most batches will have length 512, but if some samples in the dataset have length < 512, then there will occasionally be batches with length < 512 too.

By reducing the batch size for shape inference and increasing the number of batches sampled, this problem is resolved in all the cases I know about!

Reply:
What about adding a way for users to specify if the shapes are fixed or not ? Could be via a new parameter, or by checking if the feature type is `Sequence(..., length=512)`

Reply:
I think that's a good idea! We'll still need shape inference but it might be useful, and I can look into adding it when I get back!